modules/core/shared/src/main/scala/io/circe/Cursor.scala

```diff
@@ -0,0 +1,1351 @@
+              expected = JsonTypeDescription.JsonArray,
+              actual = JsonTypeDescription.fromJson(otherwise),
+              failureTarget = CursorFailureReason.FailureTarget.JsonArrayIndexTarget(n.toLong)
+            )
+          )
+      }
+    }
+
+    override final def field(k: String): Cursor = {
+      val op: CursorOp = CursorOp.Field(k)
+
+      this match {
+        case cursor: ObjectCursor =>
+          if (cursor.obj.contains(k) === false) {
+            fail(op, CursorFailureReason.NoSiblingWithMatchingKeyName(k))
+          } else {
+            ObjectCursor(cursor.obj, k, this, op, cursor.parent, cursor.changed)
+          }
+        case _: ArrayCursor =>
+          fail(op, CursorFailureReason.NoSiblingsInArrayContext(k))
+        case _: TopCursor =>
+          fail(op, CursorFailureReason.NoSiblingsInTopContext(k))
+      }
+    }
+
+    override final def downField(k: String): Cursor = {
+      val op: CursorOp = CursorOp.DownField(k)
+
+      value match {
+        case Json.JObject(obj) =>
+          if (obj.contains(k)) {
+            ObjectCursor(obj, k, this, op, this, false)
+          } else {
+            fail(op, CursorFailureReason.MissingKeyInObject(k))
+          }
+        case otherwise =>
+          fail(
+            op,
+            CursorFailureReason.IncorrectFocusTypeForNavigationOperation(
+              expected = JsonTypeDescription.JsonObject,
+              actual = JsonTypeDescription.fromJson(otherwise),
+              failureTarget = CursorFailureReason.FailureTarget.JsonObjectKeyTarget(k)
+            )
+          )
+      }
+    }
+
+    override final def failureReason: Option[CursorFailureReason] = None
+
+    private final def fail(op: CursorOp, reason: CursorFailureReason): FailedCursor =
+      FailedCursor(this, op, reason)
+  }
+
+  /** A cursor pointing to the top, i.e. root, of a JSON structure. */
+  private sealed abstract class TopCursor extends SuccessCursor {
+    override final def isRoot: Boolean = true
+
+    override final def toString: String =
+      s"TopCursor(value = $value)"
+  }
+
+  private object TopCursor {
+    private[this] final case class TopCursorImpl(
+      override val value: Json,
+      override protected val unsafeLastCursor: Cursor.SuccessCursor,
+      override protected val unsafeLastOp: CursorOp
+    ) extends TopCursor
+
+    private[Cursor] def apply(value: Json): TopCursor =
+      TopCursor(value, null, null)
+
+    private[Cursor] def apply(value: Json, unsafeLastCursor: Cursor.SuccessCursor, unsafeLastOp: CursorOp): TopCursor =
+      TopCursorImpl(value, unsafeLastCursor, unsafeLastOp)
+  }
+
+  /**
+   * An internal trait for "normal" cursors. "normal" here means not the top
+   * and not a failure state. This is useful because "normal" cursors have
+   * some common properties which it is occassionaly nice to match on.
+   */
+  private sealed abstract class NormalCursor extends SuccessCursor {
```

Comment:
`Normal` i think is a little confusing?  What other options for names could we use?  `Intermediate`? I'm not sure, but `Normal` didn't initially feel right.

Reply:
Yeah, I totally agree, but I don't know what else to call it. `Intermediate` feels equally confusing to me.

Reply:
`HasParentCursor`?

Reply:
`NotTopOrFailedCursor`? :laughing: